### ISSUE TITLE
ARM64 incorrect emulation when `UC_ERR_READ_UNMAPPED` happens

### ISSUE THREAD

**Issue by zuypt** [unicorn-engine/unicorn#1699]
How to reproduce
~~~python
from unicorn import *
from unicorn.arm64_const import *

mu = Uc(UC_ARCH_ARM64, UC_MODE_ARM)

CODE_PAGE = 0x4000

code = [
    b"\x20\x00\x80\xd2", # mov x0, #1
    b"\x41\x00\x80\xd2", # mov x1, #2
    b"\x20\x00\x40\xf9", # ldr x0, [x1]
]
code = b"".join(code)

mu.mem_map(CODE_PAGE, 1024*1024)
mu.mem_write(CODE_PAGE, code)

try:
    mu.emu_start(CODE_PAGE, CODE_PAGE + len(code))
except UcError as e:
    print ('pc: ', hex(mu.reg_read(UC_ARM64_REG_PC)))
    print ('x0: ', hex(mu.reg_read(UC_ARM64_REG_X0)))
    print ('x1: ', hex(mu.reg_read(UC_ARM64_REG_X1)))
    print (e)
~~~

After `UC_ERR_READ_UNMAPPED` pc is not updated and the value of register `x0` changed even though the read is invalid.

Expected result:
- pc is updated and not 0x4000
- x0 = 1 
- x1 = 2

Actual result:
- pc = 0x4000
- x0 = 0
- x1 = 2

**Comment by zuypt** [comment]
Is it possible to update PC after the emulation stopped or just update PC regardless of performance ?
I'm not familiar with the code base. Can you give me hint where to look at ? I would like to contribute a fix if possible.

**Comment by wtdcode** [comment]
`Is it possible to update PC after the emulation stopped or just update PC regardless of performance`

Register a `UC_HOOK_CODE` as the FAQ said.

**Comment by wtdcode** [comment]
From #1717, I remember that we should have sync-ed PC for memory read/write so this is only a bug.

**Comment by wtdcode** [comment]
I updated https://github.com/unicorn-engine/unicorn/wiki/FAQ#why-do-i-get-a-wrong-pc-after-emulation-stops and put a fix at a5d4d30a3152c7f6b7044553b530ee185e77d355

**Comment by zuypt** [comment]
Hi, Thank you for the very detailed update. The fixed python code for anyone who might need it.

~~~python
from unicorn import *
from unicorn.arm64_const import *

mu = Uc(UC_ARCH_ARM64, UC_MODE_ARM)

CODE_PAGE = 0x4000

code = [
    b"\x20\x00\x80\xd2", # mov x0, #1
    b"\x41\x00\x80\xd2", # mov x1, #2
    b"\x20\x00\x40\xf9", # ldr x0, [x1]
]
code = b"".join(code)

mu.mem_map(CODE_PAGE, 1024*1024)
mu.mem_write(CODE_PAGE, code)

def mem_read_hook(uc, access, address, size, value, data):
    return 0

mu.hook_add(UC_HOOK_MEM_READ, mem_read_hook)

try:
    mu.emu_start(CODE_PAGE, CODE_PAGE + len(code) -1)
except UcError as e:
    print ('pc: ', hex(mu.reg_read(UC_ARM64_REG_PC)))
    print ('x0: ', hex(mu.reg_read(UC_ARM64_REG_X0)))
    print ('x1: ', hex(mu.reg_read(UC_ARM64_REG_X1)))
    print (e)
~~~
